sam.c

```diff
@@ -2328,6 +2319,111 @@ int sam_parse1(kstring_t *s, sam_hdr_t *h, bam1_t *b)
     return -2;
 }
 
+static uint32_t read_ncigar(const char *in) {
+    uint32_t n_cigar = 0;
+    char *q = (char *)in;
+    for (; *q && *q != '\t'; ++q)
+        if (!isdigit_c(*q)) ++n_cigar;
+    if (!n_cigar) {
+        hts_log_error("No CIGAR operations");
+        return 0;
+    }
+    if (n_cigar >= 2147483647) {
+        hts_log_error("Too many CIGAR operations");
+        return 0;
+    }
+
+    return n_cigar;
+}
+
+/*! @function
+ @abstract  Parse a CIGAR string into preallocated a uint32_t array
+ @param  in      [in]  pointer to the source string
+ @param  a_cigar [out]  address of the destination uint32_t buffer
+ @return         number of processed input characters; 0 if error
+ */
+static int parse_cigar(const char *in, uint32_t *a_cigar, uint32_t n_cigar) {
+    int i, overflow = 0;
+    char *p, *q = (char *)in;
+    for (i = 0; i < n_cigar; i++) {
+        uint32_t len;
+        int op;
+        p = q;
+        len = hts_str2uint(q, &q, 28, &overflow)<<BAM_CIGAR_SHIFT;
+        if (q == p) {
+            hts_log_error("CIGAR length invalid at position %d (%s)", (int)(i+1), p);
+            return 0;
+        }
+        if (overflow) {
+            hts_log_error("CIGAR length too long at position %d (%.*s)", (int)(i+1), (int)(q-p+1), p);
+            return 0;
+        }
+        op = bam_cigar_table[(unsigned char)*q++];
+        if (op < 0) {
+            hts_log_error("Unrecognized CIGAR operator");
+            return 0;
+        }
+        a_cigar[i] = len;
+        a_cigar[i] |= op;
+    }
+
+    return q-in;
+}
+
+size_t sam_parse_cigar(const char *in, char **end, uint32_t **a_cigar, uint32_t *a_mem, uint32_t *a_len) {
+    size_t n_cigar = 0;
+    int diff;
+
+    if (!in || !end || !a_cigar || !a_mem || !a_len) {
+        hts_log_error("NULL pointer arguments");
+        return 0;
+    }
+    *end = (char *)in;
+
+    n_cigar = read_ncigar(in);
+    if (!n_cigar) return 0;
+    if (n_cigar > (*a_mem - *a_len)) {
+        uint32_t *a_tmp = realloc(*a_cigar, 1.5*(n_cigar + *a_len)*sizeof(**a_cigar));
```

Comment:
Given this isn't a dynamic read-and-grow cycle, I don't see the point of the 1.5*mem bit.  It's previously computed n_cigar so this should be known exactly.  I'm also not sure why it's `n_cigar + *a_len` and not just `n_cigar`?

Or is this for when called in a loop with the same buffers so we can grow to beyiond the biggest we've seen and give room before finding the next biggest?  (If so the `+ *a_len` bit looks like it'll grow and grow and grow.)  Generally though I think the htslib way for such reuse is kstring, but that doesn't seem to be a hard rule (bam1_t doesn't work that way for example).

The alternative is to do away with `read_ncigar` and keep the 1.5*mem bit so it dynamically grows as it's parsing the CIGAR string.  That makes it one pass instead of two, at the expense of potentially doing memcpy internally during reallocs.

Anyone else have views on this?

Reply:
I wanted to avoid a one pass read of the string, because that would have degraded the performance of `sam_parse1` with repeated reallocs. However, since this is a separate API method, we could look into alternative strategies.